### PR TITLE
Feature/60001 check submission api

### PR DESCRIPTION
**Pull request by GuyHarwood**
Check Submission API

- pupil SPA submits to this instead of storage queue, when configured to do so
- uses JWT
- new version 3 submitted check message schema
- ~~no string compression~~
- tech support utility for service bus queue submission
- util-submit-check updates for service bus queue dispatch

Note: refactor of `admin/services` service classes to ditch unnecessary service postfix

### CODE REVIEW

**Review by jon-shipley — COMMENTED** [review]
lgsf

File: admin/services/jwt/jwt.service.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        resolve(token ?? '')
+      })
+    })
+  }
+
+  async verify (token: string): Promise<object | string> {
+    return new Promise((resolve, reject) => {
+      jwt.verify(token, config.PupilApi.Submission.JwtSecret, (err, decoded) => {
+        if (!isNullOrUndefined(err)) { reject(err) }
+        resolve(decoded ?? '')
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/services/jwt/jwt.service.ts:22]: this looks like a 'typescript is complaining and i'm busy right now' kind of move.  will address

File: admin/tests-integration/jwt.service.spec.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  let sut
+  beforeEach(() => {
+    // create a 32 character string for a valid secret
+    const secret = '12345678901234567890123456789012'
+    config.PupilAuth.JwtSecret = secret
+    sut = JwtService.getInstance()
+  })
+
+  test('it preserves the payload', async () => {
+    const payload = { test: 'test' }
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/tests-integration/jwt.service.spec.js]: the contents of the payload are passive.  it just preserves it.
will make it more complex to prove that point.

File: tslib/src/pupil-api/controllers/submit.controller.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    const authHeader = req.headers.authorization
+    if (authHeader === undefined) {
+      return apiResponse.unauthorised(res)
+    }
+    const token = authHeader.split(' ')[1]
+    if (token == null) return apiResponse.unauthorised(res)
+    try {
+      await this.jwtService.verify(token)
+    } catch (error: any) {
+      logger.error(`JWT verification failed: ${error.message}`)
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply tslib/src/pupil-api/controllers/submit.controller.ts:39]: the body could contain literally anything, more so in the case of failed authenticity.  
Do we really want to be taking anything from the body and injecting it into our logs?
its extra work on the submission endpoint to parse the body also

i'm open minded on this - we can discuss

File: tslib/src/pupil-api/services/redis-pupil-auth.service.ts [hunk 4]
```diff
@@ -53,12 +54,14 @@ export class RedisPupilAuthenticationService implements IPupilAuthenticationServ
     const cacheKey = this.buildCacheKey(schoolPin, pupilPin)
     const preparedCheckEntry = await this.redisService.get(cacheKey)
     if (preparedCheckEntry === undefined) {
+      console.log('GUY: no prepared check')
```
Comment (jon-shipley) [inline tslib/src/pupil-api/services/redis-pupil-auth.service.ts:57]: debug?

File: tslib/src/pupil-api/services/redis-pupil-auth.service.ts [hunk 5]
```diff
@@ -53,12 +54,14 @@ export class RedisPupilAuthenticationService implements IPupilAuthenticationServ
...
       return
     }
     const pinExpiresAtUtc = R.prop('pinExpiresAtUtc', preparedCheckEntry)
     const pinValidFromUtc = R.prop('pinValidFromUtc', preparedCheckEntry)
     const currentDateTime = moment.utc()
     if (moment.utc(pinValidFromUtc).isAfter(currentDateTime) || moment.utc(pinExpiresAtUtc).isBefore(currentDateTime)) {
+      console.log('GUY: its a time validity issue')
```
Comment (jon-shipley) [inline tslib/src/pupil-api/services/redis-pupil-auth.service.ts:64]: debug?

File: tslib/src/schemas/models.ts [hunk 6]
```diff
@@ -1,10 +1,33 @@
...
+  device: Record<any, any>
+  pupil: {
+    checkCode: string
+  }
+  questions: Record<any, any>
+  school: {
+    name: string
+    uuid: string
+  }
+  tokens: Record<any, any>
```
Comment (jon-shipley) [inline tslib/src/schemas/models.ts:25]: This should be locked down precisely.

**Review by jon-shipley — COMMENTED** [review]
Epic work! 👍 😄 

I've left a few comments, as you might expect from something so huge!

File: tslib/src/functions/check-validator/validators/answer-count-check-form.validator.spec.ts [hunk 7]
```diff
@@ -25,41 +26,10 @@ describe('answer-count.validator', () => {
   beforeEach(() => {
     const checkFormServiceMock = new CheckFormServiceMock()
     jest.spyOn(checkFormServiceMock, 'getCheckFormForCheckCode').mockResolvedValue(checkForm)
+    jest.spyOn(checkFormServiceMock, 'getLiveFormQuestionCount').mockReturnValue(checkForm.length)
     sut = new AnswerCountCheckFormValidator(checkFormServiceMock)
   })
 
-  test('returns an error if the answers property is missing', async () => {
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply tslib/src/functions/check-validator/validators/answer-count-check-form.validator.spec.ts:31]: each validator does one thing, and one thing only. 
there are already validators for the check code, answers etc.. that have performed these checks, as the validators are loaded as a set and executed in sequence.

File: tslib/src/common/compression-service.ts [hunk 8]
```diff
@@ -1,28 +1,51 @@
...
     return lzString.decompressFromUTF16(input)
   }
+
+  /**
+   * Compress a string to base64
+   * @param {string}
+   * @return {*|string}
+   */
+  compressToBase64 (input: string): string {
+    console.log(`compressing input: ${input.substring(0, 30)}...`)
```
Comment (jon-shipley) [inline tslib/src/common/compression-service.ts:35]: debug?

File: tslib/src/functions-throttled/sync-results-init/sync-results-init.service.ts [hunk 9]
```diff
@@ -57,7 +57,7 @@ export class SyncResultsInitService {
     if (archive === null || archive === undefined || archive === '') {
       throw new Error(`CheckCode ${check.checkCode} has an invalid archive`)
     }
-    const payloadString = this.compressionService.decompress(archive)
+    const payloadString = this.compressionService.decompressFromUTF16(archive)
```
Comment (jon-shipley) [inline tslib/src/functions-throttled/sync-results-init/sync-results-init.service.ts:60]: Surely this is now base64 encoded (and compressed)?

File: admin/services/check-start/check-start.service.js [hunk 10]
```diff
@@ -264,11 +283,32 @@ const checkStartService = {
...
+      pupilConfig.submissionMode = config.FeatureToggles.checkSubmissionApi ? 'modern' : 'legacy'
+      let jwtToken, checkSubmissionData, checkCompleteData
+
+      if (config.FeatureToggles.checkSubmissionApi) {
+        const jwtSigningOptions = {
+          issuer: 'MTC Admin',
+          subject: o.pupil_uuid,
+          expiresIn: fiveDays // expires in 5 days
+        }
+        jwtToken = await jwtService.sign({
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/services/check-start/check-start.service.js:302]: good catch

File: docs/messaging/message-schemas.md [hunk 11]
```diff
@@ -79,6 +81,25 @@ See the [check-submission](./check-submission/readme.md) documentation for full
...
+Service bus queue for submitting checks from the pupil API.
+
+See the [check-submission](./check-submission/readme.md) documentation for full payload schema.
+
+```typescript
+{
+  version: 3
+  checkCode: string
+  schoolUUID: string
+  // TODO
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply docs/messaging/message-schemas.md:95]: yes, fixed

File: tslib/src/functions-throttled/sync-results-init/sync-results-init.service.ts (hunk 9 above)
Comment (GuyHarwood) [reply tslib/src/functions-throttled/sync-results-init/sync-results-init.service.ts]: yes, see latest commits.
that was part of a rename refactor in bulk

File: pupil-spa/gen_config.sh [hunk 12]
```diff
@@ -7,8 +7,7 @@ scriptDir="$( cd "$( dirname "$0" )" >/dev/null && pwd )"
 envFile=${scriptDir}/../.env
 [[ -f ${envFile} ]] && source ${envFile}
 applicationInsightsCode=${APPINSIGHTS_INSTRUMENTATIONKEY}
-authPingUrl=${AUTH_PING_URL:-"http://localhost:3003/ping"}
-authUrl=${AUTH_URL:-"http://localhost:3003/auth"}
+apiBaseUrl=${API_BASE_URL:-"http://localhost:3003"}
```
Comment (jon-shipley) [inline pupil-spa/gen_config.sh:10]: This might be duplication of config values.  In admin app this env var is `PUPIL_API_BASE_URL`.

Comment (GuyHarwood) [reply pupil-spa/gen_config.sh]: modified

File: tslib/src/functions/util-submit-check/index.ts [hunk 13]
```diff
@@ -33,18 +33,33 @@ const httpTrigger: AzureFunction = async function (context: Context, req: HttpRe
       numberOfDuplicateAnswers: req.body?.answerNumberOfDuplicates
     }
   }
+
+  const messageVersion: string = req.body?.messageVersion ?? SubmittedCheckVersion.V3
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply tslib/src/functions/util-submit-check/index.ts:37]: thats the idea - it just works out of the box with the 'new functionality' but anyone wanting to play an older version message can do if need be

File: admin/services/check-start/check-start.service.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/services/check-start/check-start.service.js:302]: decision on Monday

[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/services/check-start/check-start.service.js:302]: ticket created to offload JWT creation

**Review by jon-shipley — APPROVED** [review]
LGTM 👍 

Many thanks for all the changes!

**Pull request merged**